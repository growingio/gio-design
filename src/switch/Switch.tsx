import React, { useState, useEffect } from 'react';
import classnames from 'classnames';
import { get, isUndefined } from 'lodash';
import { usePrefixCls } from '@gio-design/utils';
import usePrevious from '../utils/hooks/usePrevious';
import filterChildren from '../utils/filterChildren';
import SwitchItem from './SwitchItem';
import SwitchContext from './context';
import { SwitchProps, SwitchItemValueType } from './interface';
import WithSubComponent from '../utils/withSubComponent';

const Switch = React.forwardRef<HTMLDivElement, SwitchProps>((props: SwitchProps, ref) => {
  const {
    className,
    style,
    disabled,
    defaultValue,
    value,
    onChange,
    children,
    options,
    size = 'normal',
    dataTestId = 'switch',
  } = props;

  const prefixCls = usePrefixCls('switch');
  const [selectedValue, setSelectedValue] = useState(() => (!isUndefined(value) ? value : defaultValue));
  const prevSelectedValue = usePrevious(value);

  useEffect(() => {
    if (!isUndefined(value) || value !== prevSelectedValue) {
      setSelectedValue(value);
    }
  }, [value, prevSelectedValue]);

  const getChildrenItems = () =>
    filterChildren(children, (child) => {
      if (React.isValidElement(child)) {
        if (typeof child.type !== 'object' || get(child, 'type.displayName') !== SwitchItem.displayName) {
          return false;
        }
        if (!Reflect.has(child.props, 'value')) {
          return false;
        }
        return true;
      }
      return false;
    });

  const renderItems = () => {
    let renderedChildren: React.ReactNode[] = [];
    if (options && options.length > 0) {
      renderedChildren = options
        .filter((_) => !!_)
        .map((opt) => {
          if (typeof opt === 'string') {
            return (
              <SwitchItem
                key={`${prefixCls}-option-${opt}`}
                disabled={disabled}
                value={opt}
                checked={selectedValue === opt}
              >
                {opt}
              </SwitchItem>
            );
          }

          return (
            <SwitchItem
              key={`${prefixCls}-option-${opt.value}`}
              disabled={disabled || opt.disabled}
              value={opt.value}
              checked={selectedValue === opt.value}
            >
              {opt.label}
            </SwitchItem>
          );
        });
    }

    if (children) {
      renderedChildren = [...renderedChildren, ...getChildrenItems()];
    }

    return renderedChildren;
  };

  const wrapperCls = classnames(className, `${prefixCls}__group`, {
    [`${prefixCls}__group__disabled`]: disabled,
    [`${prefixCls}__group_${size}`]: size,
  });

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (!value) {
      setSelectedValue(e.target.value);
    }
    if (onChange && e.target.value !== prevSelectedValue) {
      onChange(e);
    }
  };

  return (
    <div style={style} className={wrapperCls} ref={ref} data-testid={dataTestId}>
      <SwitchContext.Provider
        value={{
          disabled,
          value: selectedValue as SwitchItemValueType,
          onChange: handleChange,
        }}
      >
        {renderItems()}
      </SwitchContext.Provider>
    </div>
  );
});

// Switch.displayName = 'Switch';

export default WithSubComponent(Switch, {
  Item: SwitchItem,
});
