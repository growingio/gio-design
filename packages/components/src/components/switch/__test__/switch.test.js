import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Switch from '../switch.tsx';

describe('Testing Radio', () => {
  it('should mount and unmount Radio with no error.', () => {
    const wrapper = mount(<Switch />);
    expect(() => {
      wrapper.setProps({});
      wrapper.unmount();
    }).not.toThrow();
  });

  it('should render correctly', () => {
    const domTree = renderer.create(<Switch className='gio-customized-switch' disabled defaultChecked />).toJSON();
    expect(domTree).toMatchSnapshot();
  });

  it('should return "false" when change on defaultChecked Switch.', () => {
    const onChange = jest.fn();
    const wrapper = mount(<Switch defaultChecked onChange={onChange} />);
    wrapper.find('.gio-switch').at(0).simulate('click');

    expect(onChange).toHaveBeenCalled();
    // expect(onChange.mock.calls[0][0].target.checked).toBe(false);
  });

  it('switch with suffixContent', () => {
    const wrapper = mount(<Switch defaultChecked suffixContent />);
    // expect(wrapper.find('.suffixContent').text()).to.equals('开');
    expect(wrapper.exists('.suffixContent')).toBe(true);
    expect(wrapper.find('.suffixContent').text()).toEqual('开');
  });
});
