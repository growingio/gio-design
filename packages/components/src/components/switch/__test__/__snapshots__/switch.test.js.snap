// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Radio should render correctly 1`] = `
<div>
  <div
    className="gio-switch gio-switch-checked gio-switch-disabled gio-customized-switch"
    onClick={[Function]}
    style={
      Object {
        "background": undefined,
        "borderColor": undefined,
      }
    }
  >
    <div
      className="gio-switch-handle"
      style={
        Object {
          "background": undefined,
        }
      }
    />
  </div>
</div>
`;
